### PR TITLE
Hyperparameter Tuning of Machine Learning Model in Python

### LINKED ISSUE #5009: [Machine Learning] Hyperparameter Tuning of Machine Learning Model in Python
#### Proposed title of article
[Machine Learning] Hyperparameter Tuning of Machine Learning Model in Python

#### Proposed article introduction
Hyperparameters are parameters that can be fine-tuned and adjusted in order to improve the performance of a machine learning model. Various machine algorithms such as RandomForest and K-nearest neighbor algorithms have parameters that can be adjusted to achieve an optimized model.

In the process of hyperparameter tuning, we create a list of all possible values for hyperparameters.  We then iterate through the list trying to find a random combination of hyperparameters, record the performance of each parameter, and finally return the combination of hyperparameters that will provide an optimal model with the optimal solution.

Finding and tuning the right hyperparameters for any machine learning model is a very crucial aspect of today's development. We have to reduce the model loss function and in the end, we have a better model. We have various techniques for hyperparameter tuning: Random Search, Grid Search, Bayesian Optimization, and Tree-structured Parzen estimators (TPE).

#### Key takeaways 
1. What is hyperparameter tuning.
2. Various techniques and methods used for hyperparameter tuning.
3. Building a machine learning model using RandomForest.
4. Fine-tuning RandomForest parameters.
5. Implementing various techniques and methods for hyperparameter tuning.

#### Article quality
This tutorial is a practical guide, explaining everything from scratch with a step-by-step guide that is easier to follow. This tutorial also uses two approaches in building the model, In the First approach we build a model without tuning and in the second approach, we will fine-tune the parameters to increase the model accuracy.
The tutorial will also discuss the methods for hyperparameter tuning in detail.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by willyngashu**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5009

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by LinusMuema** [comment]
Hi @willyngashu 

Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content. Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[Hyperparameter Tuning of Machine Learning Model in Python #5163 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7608449/Hyperparameter.Tuning.of.Machine.Learning.Model.in.Python.5163.Copyleaks.Report.pdf)

**Comment by willyngashu** [comment]
Hello @Bayler, I have made changes according to the plagiarism check. However, some are imports for common machine learning packages which cannot be edited. Kindly review.
cc @LinusMuema.

**Comment by hectorkambow** [comment]
@Bayler 
Why does this need another plagiarism check??

**Comment by Bayler** [comment]
> @Bayler 
> Why does this need another plagiarism check??

@hectorkambow The article was flagged for plagiarism. I thought after the author has made the revisions, we perform another check to ensure everything's okay now

**Comment by hectorkambow** [comment]
No this was fine. 👍  cc @LinusMuema  

the % was .9 over 10% that is perfectly fine. 

we close PRs that are near or close to 30% - we ask students to make corrections if it is higher than 10% - the higher (and closer to 30%) then we may perform another check. 

This .9 edits does not require another check. 

cc @Bayler

**Comment by hectorkambow** [comment]
cc @section-engineering-education/content-moderator-team see my comment about running multiple plagiarism reports.

File: content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+#### Grid Search
+This is a brute force searching technique. In this technique, we create a list of all the combination values for hyperparameters.  We then iterate through all hyperparameters. Finally, it records the best performing hyperparameters used in model training. This is shown below.
+
+![Grid Search Technique](/engineering-education/hyperparameter-tuning-of-machine-learning-model-in-python/grid-search-technique.png)
+
+*[Image source: Medium](https://miro.medium.com/max/534/1*6mLLIGYLny9B6GAnVZcltw.png)*
+
+#### Random Search
+In this technique, we also create a list of all the combination values for hyperparameters. It's similar to grid search, but uses random search instead of using exhaustive search.
```
Comment (Bayler) [inline content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md:59]:
```suggestion
We also create a list of all the combination values for hyperparameters in this technique. It's similar to grid search, but it uses random search instead of exhaustive search.
```

File: content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+#### Bayesian Optimization
+This technique uses probability to find a model with the minimum loss function. It does this by mapping the hyperparameters to the function that will produce an optimal model.
+Bayesian Optimization ensures that the process takes the minimum number of steps.
+
+#### Gradient-based optimization
+It is best used with the [gradient descent algorithm](/engineering-education/gradient-descent-algorithm-in-python/). It fine-tunes the parameters for the gradient descent algorithm to produce an optimal model.
+
+#### Evolutionary optimization
+This technique uses the concept of natural selection in hyperparameter tuning. They use the evolution process and survival of the fittest by [Charles Darwin](https://en.wikipedia.org/wiki/Charles_Darwin).
```
Comment (Bayler) [inline content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md:74]: They use....it is unclear who or what They refers to

File: content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+This is target output, this can either be a `1` or `0`. This is prediction output of the model.
+
+- `n_features=10`: 
+These are the independent variables that are used as input for the model. The model will have a total of `10` input columns
+
+- `n_redundant=0`:
+This specifies the number of repeated data points in the dataset.
+
+- `random_state=1`
+Is used to set the seeding factor used to randomly generate our dataset. This ensures that the model results can be reproduced and applied elsewhere.
```
Comment (Bayler) [inline content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md:109]:
```suggestion
It is used to set the seeding factor used to generate our dataset randomly. This ensures that the model results can be reproduced and applied elsewhere.
```

File: content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+Let's now start tuning the hyperparameters. 
+
+To perform hyperparameter tuning, we have to specify the range `max_features` and `n_estimators`. These will be used to create a grid of hyperparameters.
+
+We specify the range using `NumPy`. Let's import `NumPy` using the following code.
+
+```python
+import numpy as np
+```
+Let's now create a range of `max_features` and `n_estimators`.
```
Comment (Bayler) [inline content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md:259]:
```suggestion
Now, let's create a range of `max_features` and `n_estimators`.
```

File: content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+```python
+n_estimators_range = np.arange(10,210,10)
+```
+The output is shown below.
+
+![n_estimators_range](/engineering-education/hyperparameter-tuning-of-machine-learning-model-in-python/n-estimators-range.png)
+
+The range of `n_estimators` will be between `10` and `200`.
+
+Let's now use `max_features` and `n_estimators` to build our grid.
```
Comment (Bayler) [inline content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md:279]: Now, let's use `max_features` and `n_estimators` to build our grid.

File: content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+```
+
+The `param_grid` uses the `max_features=max_features_range` and `n_estimators=n_estimators_range` as the input.
+
+We now initialize the algorithm we want to fine-tune. We want to finetune the `RandomForestClassifier()` algorithm.
+
+```python
+rf = RandomForestClassifier()
+```
+Now that we have initialized the algorithm, let now initialize the `GridSearchCV` function.
```
Comment (Bayler) [inline content/articles/hyperparameter-tuning-of-machine-learning-model-in-python/index.md:295]: The use of the second "now" is unnecessary.

**Comment by Bayler** [comment]
@willyngashu Great read! Please work on the minor edits before we can proceed for final review.

**Comment by willyngashu** [comment]
Thanks, @Bayler, for the review. I have implemented the changes and the necessary changes.

**Comment by Bayler** [comment]
> Thanks, @Bayler, for the review. I have implemented the changes and the necessary changes.

Please take note of the commits I've made to improve the flow of the tutorial. Otherwise, great tutorial!

**Comment by ahmadmardeni1** [comment]
Great work gentlemen @willyngashu  @Bayler 
Please go over the changes I've made.
